File: LC/python/Path Sum II.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @param sum, an integer
    # @return a list of lists of integers
    def pathSum(self, root, sum):
        def dfs(root,tmp,string):
            if root == None:
                return
            string.append(root.val)
            tmp += root.val
            if root.left == None and root.right == None:
                if tmp == sum:
                    res.append(list(string))
            else:
                dfs(root.left,tmp,string)
                dfs(root.right,tmp,string)
            string.pop()
        
        if root == None:
            return []
        res = []
        dfs(root,0,[])
        return res

File: LC/java/Path Sum II.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ArrayList<ArrayList<Integer>> pathSum(TreeNode root, int sum) {
        ArrayList<Integer> sub = new ArrayList<Integer>();
        ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        
        if(root == null) return res;
        
        helper(root,sum,0,sub,res);
        return res;
    }
    
    private void helper(TreeNode root, int sum, int curr, ArrayList<Integer> sub, ArrayList<ArrayList<Integer>> res){
        if(root==null) return;
        
        if(root.left==null && root.right==null && curr+root.val == sum){
            sub.add(root.val);
            res.add(new ArrayList<Integer>(sub));
            sub.remove(sub.size()-1);
        }
        
        sub.add(root.val);
        helper(root.left, sum, root.val+curr, sub, res);
        helper(root.right, sum, root.val+curr, sub, res);
        sub.remove(sub.size()-1);
    }
}